libs/common/ring_buffer.hpp

```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef IROHA_COMMON_RING_BUFFER_HPP
+#define IROHA_COMMON_RING_BUFFER_HPP
+
+namespace iroha {
+
+    template<typename T, size_t Count> class RingBuffer final {
+    public:
+        using type = T;
+        using handle = size_t;
+        using callback = std::function<void(handle, T&)>;
+
+    private:
+        static_assert(Count > 0, "Unexpected count value. It must be above 0.");
+        enum { kSizeType = sizeof(type) };
+
+        struct Node {
+            uint8_t data[kSizeType];
+            bool active;
+        } data_[Count];
+        handle last_;
+
+        inline void setActive_(Node& node, bool active) {
+            node.active = active;
+        }
+
+        inline bool isActive_(Node const& node) const {
+            return node.active;
+        }
+
+        inline type& getItem_(Node& node) {
+            assert(isActive_(node));
+            return *(type*)node.data;
+        }
+
+        inline type const& getItem_(Node const& node) const {
+            assert(isActive_(node));
+            return *(type const*)node.data;
+        }
+
+        inline void destruct_(Node& node) {
+            assert(isActive_(node));
+            auto& item = getItem_(node);
+
+            setActive_(node, false);
+            item.~type();
+        }
+
+        template<typename...__args> inline void construct_(Node& node, __args&&... args) {
+            assert(!isActive_(node));
+
+            new(node.data) type(std::forward<__args>(args)...);
+            setActive_(node, true);
+        }
+
+        inline Node& toNode_(handle h) {
+            assert(h < Count);
+
+            auto& node = *(Node*)&data_[h];
+            return node;
+        }
+
+        inline Node const& toNode_(handle h) const{
+            assert(h < Count);
+
+            auto const& node = *(Node const*)&data_[h];
+            return node;
+        }
+
+        inline void store_(handle h) {
+            last_ = h;
+        }
+
+        inline handle load_() {
+            return last_;
+        }
+
+        inline handle iterateNext_(handle h) {
+            return ++h % Count;
+        }
+
+        inline handle iteratePrev_(handle h) {
+            return (h + (Count - 1)) % Count;
+        }
+
+        inline bool empty_() {
+            return !isActive_(toNode_(last_));
+        }
+
+    public:
+        RingBuffer() : last_(0) {
+            memset(data_, 0, sizeof(data_));
+        }
+
+        ~RingBuffer() {
+            while (!empty_())
+                pop();
+        }
+
+        template<typename...__args>
+        void push(callback&& on_add, callback&& on_remove, __args&&... args) {
```

Comment:
The difference will be only in inlining, because gcc will not inline it. I remains std::function to use it in using, to show the class user the signature of called function.

0x555555662ed6  mov    rsi,r13
0x555555662ed9  mov    WORD PTR [rbp-0xbf0],cx
0x555555662ee0  mov    DWORD PTR [rbp-0xc1b],0x0
0x555555662eea  mov    BYTE PTR [rbp-0xc17],0x1
0x555555662ef1  mov    QWORD PTR [rbp-0xbe8],0x1
0x555555662efc  mov    QWORD PTR [rbp-0xa40],0x1
0x555555662f07  mov    QWORD PTR [rbp-0xd18],rax
0x555555662f0e  movaps XMMWORD PTR [rbp-0x830],xmm0
0x555555662f15  call   0x55555567dc30 <std::_Function_handler<void (unsigned long, Test&), main::{lambda(unsigned long, Test&)#1}>::_M_invoke(std::_Any_data const&, unsigned long&&, Test&)>
0x555555662f1a  mov    rax,QWORD PTR [rbp-0x830]
0x555555662f21  test   rax,rax
0x555555662f24  je     0x555555662f33 <main+243>

There are no additional calls due to std::function.

but msvc will^)
GCC will inline it with __always_inline.